### DISCUSSION TITLE
How to configure a module based on environment variables

### DISCUSSION THREAD

**Discussion by mlbiche** [nuxt/nuxt#24905] [Questions] [answered]
I am looking for a way to use environment variables to configure some modules.

The only related resource I have found is this discussion that finally points to a solution avoiding using env vars in nuxt.config.js : https://github.com/nuxt/nuxt/discussions/15931#discussioncomment-3898711

What I want to achieve is to set Content-Security-Polcy in the `nuxt-security` module using an env vars.

In my .env file, I have :

```bash
MATOMO_HOST=...
```

which is set in my repository pipelines and Docker container with the variable `NUXT_PUBLIC_MATOMO_HOST`. And here is what I am trying the nuxt.config.js file :

```ts
export default defineNuxtConfig({
  // ...
  runtimeConfig: {
    public: {
      matomo: {
        host: process.env.MATOMO_HOST,
      },
    }
  },
  // ...
  security: {
    headers: {
      contentSecurityPolicy: {
        'connect-src': [
          "'self'",
          process.env.MATOMO_HOST ||
            'https://my.matomo.host',
        ]
      }
    }
  }
})
```

It is working locally, either in dev mode or building the app and running it locally, but the CSP always contains `https://my.matomo.host` and not the environment value when deploying the application. While the runtime config is properly set.

I have also tried using `process.env.NUXT_PUBLIC_MATOMO_HOST` or `import.meta.env.NUXT_PUBLIC_MATOMO_HOST` without any success.

**Comment by TheAlexLichter** [comment]
You'd have to set everything through the `runtimeConfig` and pick the matching env variable. Env variables used in the `nuxt.config.ts` will only be applied during *build*. I've [talked about that recently in a video](https://www.youtube.com/watch?v=_FYV5WfiWvs).

For `nuxt-security`, there are discussions which makes it the easiest to adapt headers during runtime. See e.g. https://github.com/Baroshem/nuxt-security/issues/258 and https://github.com/Baroshem/nuxt-security/pull/298
There, the idea is to use custom hooks (which you could fill via env variables again).

**Reply by mlbiche** [reply]
Thank you @manniL for your quick answer and the instructive video.

I have also checked a bit the `nuxt-security` discussions but I am not sure that it fits my use case (or i am misundersting the discussions).

I do not aim to apply the CSP value per request by accessing the runtime config in a middleware or so, but actually have my general configuration based on such value.

Maybe this PR is closer to what I want : https://github.com/Baroshem/nuxt-security/pull/233.

But I am surprised that the proposed solution is specific to the `nuxt-security` module while it would be interesting to be able to set module config based on env variables, or any other config (such as `head` as discussed in the mentioned comment above).

**Reply by TheAlexLichter** [reply, marked as answer]
Yes, either (my or your linked PR) would work for this ☺️

> But I am surprised that the proposed solution is specific to the nuxt-security module while it would be interesting to be able to set module config based on env variables, or any other config (such as head as discussed in the mentioned comment above).

Usually, modules expose their runtimeConfig options. See e.g. the [i18n-module](https://i18n.nuxtjs.org/options/runtime-config)
